File: Ruby/stack.rb
class Cell
  :@m_value
  :@m_next

  def initialize value, n
    @m_value = value
    @m_next = n
  end

  def value
    @m_value
  end

  def next
    @m_next
  end

  public def to_s
    @m_value
  end
end

class Stack
  @m_first
  @m_size

  def initialize
    @m_first = nil
    @m_size = 0
  end

  def push value
    @m_first = Cell.new(value, @m_first)
    @m_size = @m_size + 1
  end

  def top
    if @m_first != nil
      @m_first.value
    else
      "Stack Empty"
    end
  end

  def pop
    if @m_first != nil
      firstvalue = @m_first.value
      @m_first = @m_first.next
      @m_size = @m_size - 1
      firstvalue
    else
      "Stack Empty"
    end
  end

  def size
    @m_size
  end

  def to_s
    "[#{to_sX @m_first, true}]"
  end

  private def to_sX current, first
    if current != nil
      "#{first ? "" : ","}#{current.value}#{to_sX current.next, false}"
    else
      ""
    end
  end
end

s = Stack.new
puts s
puts "Size: #{s.size}"
puts

s.push 1
puts s
puts "Top: #{s.top}"
puts "Size: #{s.size}"
puts

s.push 2
puts s
puts "Top: #{s.top}"
puts "Size: #{s.size}"
puts

s.push 3
puts s
puts "Top: #{s.top}"
puts "Size: #{s.size}"
puts

puts "Pop: #{s.pop}"
puts s
puts "Top: #{s.top}"
puts "Size: #{s.size}"
puts

puts "Pop: #{s.pop}"
puts s
puts "Top: #{s.top}"
puts "Size: #{s.size}"
puts

puts "Pop: #{s.pop}"
puts s
puts "Size: #{s.size}"
puts

puts "Top: #{s.top}"
puts "Pop: #{s.pop}"

File: lua/stack.lua
Class = {}
Class.new = function(self, object)
  object = object or {}
  setmetatable(object, self)
  self.__index = self
  return object
end

Cell = {}
Cell.construct = function(self, value, next)
  return Class.new(self, {m_value = value, m_next = next})
end

Cell.value = function(self)
  return self.m_value
end

Cell.next = function(self)
  return self.m_next
end

Cell.__tostring = function(self)
  return tostring(self.m_value)
end

Stack = {}
Stack.construct = function(self)
  return Class.new(self, {m_first = nil, m_size = 0})
end

Stack.push = function(self, value)
  self.m_first = Cell:construct(value, self.m_first)
end

Stack.top = function(self)
  assert(self.m_first ~= nil, "Top: Stack Empty.")
  return self.m_first:value()
end

Stack.pop = function(self)
  assert(self.m_first ~= nil, "Pop: Stack Empty.")
  top_value = self.m_first:value()
  self.m_first = self.m_first:next()
  return top_value
end

Stack.size = function(self)
  return self.m_size
end

Stack.__tostring = function(self)
  return "[" .. Stack.__tostringX(self.m_first, true) .. "]"
end

Stack.__tostringX = function(current, first)
  if current ~= nil then
    return ((first and "") or ",") .. current:value()
           .. Stack.__tostringX(current:next(), false)
  else
    return ""
  end
end

s = Stack:construct()
print(s)
print("Size: " .. s:size())
print()

s:push(1)
print(s)
print("Top: " .. s:top())
print("Size: " .. s:size())
print()

s:push(2)
print(s)
print("Top: " .. s:top())
print("Size: " .. s:size())
print()

s:push(3)
print(s)
print("Top: " .. s:top())
print("Size: " .. s:size())
print()

print("Pop: " .. s:pop())
print(s)
print("Top: " .. s:top())
print("Size: " .. s:size())
print()

print("Pop: " .. s:pop())
print(s)
print("Top: " .. s:top())
print("Size: " .. s:size())
print()

print("Pop: " .. s:pop())
print(s)
print("Size: " .. s:size())
--print("Top: " .. s:top())
--print("Pop: " .. s:pop())
